### ISSUE TITLE
Can't connect because "404: Not Found"

### ISSUE THREAD

**Issue by mkurz** [schlatterbeck/snxvpn#7]
Hi @schlatterbeck.

it's really great you work on this. I got `snxvpn` installed and running on my Ubuntu 16.10 laptop - but not getting a connection so far.

So here is my story, maybe (and hopefully ;) you can help me. My problem is I am not really into VPN and how all this stuff works.
Currently I am doing a project for a company and to get into their VPN I have to use the Checkpoint VPN client which only works on Windows. So I installed the `Check Point  Endpoint Security VPN` Client in a Windows 10 Virtual Machine. So all I have to do is enter the IP address of the host, my username and my password. That's it. I am connected to the VPN. No certificate to supply. (The authentication method in the VPN options is just "username and password").

I installed the `snx` application by downloading it directly from the vpn host in the browser.
![1](https://user-images.githubusercontent.com/644927/28174424-c366f78c-67f1-11e7-811b-20a6e4efa968.png)

When running `snxconnect` I now get:
```
$ snxconnect -H <ipaddress> -U <user>
Traceback (most recent call last):
  File "/home/mkurz/.local/bin/snxconnect", line 5, in <module>
    main ()
  File "/home/mkurz/.local/lib/python2.7/site-packages/snxconnect.py", line 511, in main
    result = rq.login ()
  File "/home/mkurz/.local/lib/python2.7/site-packages/snxconnect.py", line 168, in login
    self.open ()
  File "/home/mkurz/.local/lib/python2.7/site-packages/snxconnect.py", line 246, in open
    self.f = f = self.opener.open (rq, timeout = 10)
  File "/usr/lib/python2.7/urllib2.py", line 435, in open
    response = meth(req, response)
  File "/usr/lib/python2.7/urllib2.py", line 548, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python2.7/urllib2.py", line 473, in error
    return self._call_chain(*args)
  File "/usr/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 556, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
urllib2.HTTPError: HTTP Error 404: Not Found
```
So it looks like `sslvpn/Login/Login` doesn't exist. So I tried `-F`:
```
$ snxconnect -H <ipaddress> -F "" -U <user>

https://xxx.xx.xx.xxx/
No RSA javascript file found, cannot login
```

Do you have an idea what the problem could be here?
If you need the IP address of the VPN I could send it to you via email so you could investigate yourself (?)

Or do you have another idea how to connect to this VPN? I heard that it may work with https://www.shrew.net/ but also I couldn't get it working because it's quite complicated to set up. Also there you have to supply a certificate which I don't have (and maybe also not get from the sysadmins of the VPN). (Or do you have an idea how to set shrew up and get the certificate?)

**Comment by mkurz** [comment]
For completion here are the screenshots of the application in Windows.
Again, all I have is an **IP address, username and the password** - nothing more.
![5](https://user-images.githubusercontent.com/644927/28175313-2e1feb72-67f4-11e7-8fdf-a23586a64446.png)
![1](https://user-images.githubusercontent.com/644927/28175337-3d7ab386-67f4-11e7-8d00-f7e88d51f382.png)
![3](https://user-images.githubusercontent.com/644927/28175345-44eef942-67f4-11e7-9506-617bb6e44040.png)
![4](https://user-images.githubusercontent.com/644927/28175353-4b338f66-67f4-11e7-938f-006c2df8874d.png)
![2](https://user-images.githubusercontent.com/644927/28175368-51b04500-67f4-11e7-80b9-33c3020d3206.png)

**Comment by schlatterbeck** [comment]
Can you check if the browser displays an URL when you connect to this IP-Address? The first screenshot where you have downloaded the snx program should have the base-URL you have to specify with -F. There is probably some path after the IP-Address in the browser.

That said: It will probably work with snx alone: The snx program should be able to connect with just username and password: Put the following lines into your .snxrc (it is .snxrc not .snxvpnrc):

server: ip-address
username: your-snx-username

replacing the values with your known ip-address and username.
Then you should be able to connect by just calling "snx" without parameters.

Let me know if it works for you.

Note that with the appropriate Java Plugin in your browser you *should* be able to connect via the browser. I've found that this doesn't work (for me) for newer browsers though.

**Comment by mkurz** [comment]
Thanks for your reply.
I tried `snx` itself and that's what I got:
```
$ snx
Check Point's Linux SNX
build 800007116
Please enter your password:
SNX: SSL Network Extender gateway is not available. Try to connect later.
```
I always get and got this message, a few weeks back already. Two minutes later and before however I was able to connect via the windows client.
However if I provide a wrong password I get `SNX: Authentication failed` - so it looks like authorization is checked but than something fails in the next step(s)....
Is it possible that somehow access via snx was disabled server side? Because I can always connect via the windows client shown in the screenshots above...
Are the Linux `snx` client and the Windows client two different things - I mean do they use two different protocols? Is it possible that the admins disabled access via snx/web/java-plugin?

About the base url: After the IP address there is nothing more. I get the page shown in the first screenshot just by entering the ip address. No path, just `/`.
But there is also an popup coming up with the url `https://xxx.xxx.xxx.xxx/extender.html`:
![extender](https://user-images.githubusercontent.com/644927/28407065-13ebe5d2-6d33-11e7-8f2a-4ec42b63e5dc.png)


If you have any idea how to get this working I would be really greatful.

**Comment by mkurz** [comment]
More information: I now ran `snx` with `debug 5` in the `.snxrc` file and a `snx.elg` log file was created.
Here are the interesting parts:

It looks like there was a connection established (at least for authentication - I am not sure if there was a VPN tunnel created already though):
```
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] ssl_link_fwasync_client_handler: server_fingerprint_for_verification is NULL, connection was done from command line.
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] ssl_tunnel::link_established_cb: link is established - starting authentication and Office Mode
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] send_hello_to_server: message is:
(client_hello
	:client_version (1)
	:protocol_version (1)
	:protocol_minor_version (1)
	:OM (
		:ipaddr (0.0.0.0)
		:keep_address (false)
	)
)
```

Then I get the disconnect message:
```
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] ssl_link_fwasync_client_handler: state: RECEIVE_PACKT - entering
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] ssl_link_fwasync_client_handler: RECEIVE_PACKT state - expecting 111 bytes
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] ssl_link:: handle_received_packet: receiving packet: length=111, type= 1
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] ssl_tunnel::packet_receive_cb: Slim Protocol packet arrived
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] SetVerifier::add: Add handler for attribute code
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] SetVerifier::add: Add handler for attribute message
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] SetVerifier::check: Attribute code seen for the first time
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] SetVerifier::check: Attribute message seen for the first time
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] parse_disconnect_msg: Disconnecting slim with message (disconnect
	:code (205)
	:message ("SSL Network Extender gateway is not available. Try to connect later.")
)
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] ssl_tunnel::slim_disconnected: ppp is down - close tunnel
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] ssl_link_fwasync_client_handler: received packet ok
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] fwasync_mux_in: 6: rc=1, next: 80f70b0 with 3, req: 8r, 0w
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] fwasync_mux_in: 6: got 0 of 8 bytes == 8 bytes required
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] fwasync_mux_in: 6: peer closed connection
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] fwasync_end_conn: scheduling the end of connection 6
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] tunnel_stop_handler: called!
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] ssl_link:: ~ssl_link: delete link
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] ssl_link:: ssl_link_fwasync_end_handler: ending connection 
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] ssl_tunnel::link_failure_cb: got link failure, close tunnel
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] ssl_tunnel::link_failure_cb: scheduling reuse to 2000 milli-seconds
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] ssl_tunnel::tunnel_stop: error: SSL Network Extender gateway is not available. Try to connect later.
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] snx_mngr:: tunnel_down_cb: tunnel closed - SSL Network Extender gateway is not available. Try to connect later.
[ 12666 -148270976]@mkurz-notebook[20 Jul 10:28:12] snx_mngr:: tunnel_down_cb:  quit.
```

**Comment by mkurz** [comment]
OK I think I understood one thing about authentication: I was able to authenticate via username/password via the HTML formular (happens in the logs before the lines I copied/pasted above) but then the "real" authentication for the VPN doesn't work (`ssl_tunnel::link_established_cb: link is established - starting authentication and Office Mode`)...

**Comment by schlatterbeck** [comment]
First of all: From the screenshot above, the crossed-out https looks really fishy. Are you sure you're connecting via https? If yes, and this is Chrome's way to display a certificate problem, you should definitely get the certificate of your peer's TLS root-certificate into your trusted certificate store. I guess this is also the reason why snx fails to connect. The link_failure errors when the peer closes the connection might point in that direction, too.
That said: From the reverse-engineering of the snx protocol I know that the peer *can* distinguish different clients (e.g. the Active-X control used by Windows and the Java stuff used by Linux). So it might be, however unlikely, that access from Linux was selectively disabled. Do you know of other users of this VPN if they are using alternative operating systems (e.g. Apple)?

Your next step should be to get the root-cert (even if it is self-signed) of the VPN gateway into your trusted certificates and retry. You should verify that in the browser you see a proper tls connection (without crossed-out https etc.).

HTH
Ralf

**Comment by mkurz** [comment]
I finally solved this problem by using the open source [`Shrew Soft VPN Client`](https://www.shrew.net/) instead of `snx`. Now I am able to connect to the VPN. :tada: 

Maybe you (and others) are interested in how this works so I am going to share the setup:

First of all I didn't have the server certificate needed because the admins didn't provide it to me.
(In my case this is a different certificate than the one of `https://<ip>` which you could just export from the browser or via the `openssl` command line tool).

So how to get the certificate? [This awesome blog post](http://checkpoint-master-architect.blogspot.co.at/2012/09/e75x-endpoint-connect-client-decrypting.html) gave me all the hints I needed:
It turns out the certificate needed is stored in the **encrypted** `trac.config` file under `C:\Program Files (x86)\CheckPoint\Endpoint Connect` of the Windows client. To decrypt it you need to quit the GUI *and* also shutdown the `Check Point Endpoint Security VPN` *windows service*.
Afterwards edit the `trac.defaults` file (not the `trac.config`) and set the `OBSCURE_FILE` paramater from 1 to 0. Start the windows service and afterwards the Client GUI again. Now the `trac.config` is readable!

In this file you will find the certificate in the `ica_cert_object` attribute of a `<PARAM></PARAM>` tag.
copy this certificate (preserving the line breaks) into a own file and call it something like `vpn_certificate.crt`. Make sure you surround the lines with `-----BEGIN/END CERTIFICATE-----` so it looks like this:

```
-----BEGIN CERTIFICATE-----
<certificate lines from the decrypted trac.config file>
-----END CERTIFICATE-----
```

So now we have the certificate we need and can go ahead and install and setup the `Shrew Soft VPN Client`:
```
sudo apt-get install ike
sudo apt-get install ike-qtgui # GUI to make setup easier
```
Now you can setup a VPN Connection via the shrew GUI (`qikea`) like described in the [Client Configuration](https://www.shrew.net/support/Howto_Checkpoint#Client_Configuration) section of the Shrew Soft Wiki - by using the `vpn_certificate.crt` file as certificate.
(These resources also helped me setting up the Shrew VPN Client: [1](https://lists.shrew.net/pipermail/vpn-help/2010-April/007087.html) [2](https://lists.shrew.net/pipermail/vpn-help/2010-April/007088.html))
You probably also have to enter the ip ranges found under the `<IP_RANGE></IP_RANGE>` tag of the `trac.config`. These pages helped my to convert the IP ranges to CIDR Subnet Mask Notation: [1](https://ip2cidr.com/bulk-ip-to-cidr-converter.php) and [2](http://doc.m0n0.ch/quickstartpc/intro-CIDR.html).

You may also want to use [this example config](https://gist.githubusercontent.com/mkurz/4e4f8e941d03948ad579085759c4da01/raw/4d386350789f0c5038030a3b91ca289ccb40d0c7/Example-Config.vpn) which you can import via the GUI. Just save it as `Example-Config.vpn`, import it and change the ip address of the `Remote host`, supply your certificate under `Authentication` -> `Credentials` and change the IP ranges under `Policy`.

After you have set up the VPN connection you connect via the GUI or call `ikec -r "My-VPN" (-u <user> -p <password> -a)` (or for a GUI use `qikec` with the same parameters). (On the command line press `h` to display the help, which tells you do press `d` to disconnect or `q` to quit, or `s` for status).

To see whats going on and debug the client have a look into `/var/log/iked.log`. You can set the `log_level` to `loud` or `decode` in `/etc/iked.conf` - after which you problably have to restart the `iked` daemon.

If you don't want to install the GUI (or can't because you are connected to a machine via ssh) you probably want to know that the VPN connections are stored in the `$HOME/.ike/` folder. So you can just install `ike` on a server, copy the configs over and connect.

**Comment by mkurz** [comment]
One more thing:
I never got `snx` to work for me. Probably you are right and there is a certificate problem with `snx`, however even after I imported the certificate into my Ubuntu installation (like I described [here](https://github.com/schlatterbeck/snxvpn/issues/6#issuecomment-317659570)) it
still doesn't work. I have no idea which certificates `snx` is using or where it picks up the certificates from to connect or how to add own certificates.

About `snxconnect:` The HTML DOM you parse and my server provides look completely different, e.g. the elements you are looking via id are not present, etc. I tried to hack your script but it just took too much time to customize it for my needs. And because the Shrew Soft client is working for me now I don't have the need to hack into this further.
